lib/vuln/csp_validator/engine.py

```diff
@@ -0,0 +1,189 @@
+    if target_type(target) != 'SINGLE_IPv4' or target_type(target) != 'DOMAIN' or target_type(target) != 'HTTP':
+        # requirements check
+        new_extra_requirements = extra_requirements_dict()
+        if methods_args is not None:
+            for extra_requirement in extra_requirements_dict():
+                if extra_requirement in methods_args:
+                    new_extra_requirements[
+                        extra_requirement] = methods_args[extra_requirement]
+        extra_requirements = new_extra_requirements
+        if ports is None:
+            ports = extra_requirements["csp_validator_ports"]
+        if target_type(target) == 'HTTP':
+            target = target_to_host(target)
+        threads = []
+        total_req = len(ports)
+        # print(total_req)
+        thread_tmp_filename = '{}/tmp/thread_tmp_'.format(load_file_path()) + ''.join(
+            random.choice(string.ascii_letters + string.digits) for _ in range(20))
+        __log_into_file(thread_tmp_filename, 'w', '1', language)
+        trying = 0
+        keyboard_interrupt_flag = False
+        for port in ports:
+            port = int(port)
+            t = threading.Thread(target=__csp_validator,
+                                 args=(target, port, timeout_sec, log_in_file, language, time_sleep,
+              thread_tmp_filename, socks_proxy, scan_id, scan_cmd))
+            threads.append(t)
+            t.start()
+            trying += 1
+            if verbose_level > 3:
+                info(
+                    messages(language, "trying_message").format(trying, total_req, num, total, target, port, 'csp_validator_vuln'))
+            while 1:
+                try:
+                    if threading.activeCount() >= thread_number:
+                        time.sleep(0.01)
+                    else:
+                        break
+                except KeyboardInterrupt:
+                    keyboard_interrupt_flag = True
+                    break
+            if keyboard_interrupt_flag:
+                break
+        # wait for threads
+        kill_switch = 0
+        kill_time = int(
+            timeout_sec / 0.1) if int(timeout_sec / 0.1) is not 0 else 1
+        while 1:
+            time.sleep(0.1)
+            kill_switch += 1
+            try:
+                if threading.activeCount() is 1 or kill_switch is kill_time:
+                    break
+            except KeyboardInterrupt:
+                break
+        subs = __csp_validator(target, port, timeout_sec, log_in_file, language, time_sleep,
+              thread_tmp_filename, socks_proxy, scan_id, scan_cmd)
+        if len(subs) is not 0:
+            info(messages(language, "len_subdomain_found").format(len(subs)))
+            for sub in subs:
+               
+                data = json.dumps({'HOST': target, 'USERNAME': '', 'PASSWORD': '', 'PORT': '', 'TYPE': 'CSP domains found!!',
```

Comment:
this is incorrect: TYPE': 'CSP domains found!!', should be the module/finding identifier e.g. 'csp_validator_scan'  - there is inconsistent use of module id throughout this proposed module e.g. 'content_security_policy_vuln', 'csp_validator_vuln' ,'csp_validator_scan' etc. 

The idea of the module is good, but implementation needs some work. Also no actual CSP vulnerability/policy validation actually happens - only subdomain extraction, if that was the intention of this module then it needs to be renames to something like 'csp_domain_extraction_scan'

Reply:
Thank you @securestep9 for the suggestions. As per suggested, then this module should fall in /lib/scan directory instead of /lib/vuln. I moved it to scan. Also, I changed the module name to csp_domain_extraction_scan. Please review it and suggest any changes if any